echo_server.py

```diff
@@ -50,22 +52,45 @@ def response_error(error_num, error_msg):
 
 def parse_request(request):
     request_list = request.split()
-    print len(request_list)
-    if len(request_list) != 5:
-        return None
     error_check = check_for_errors(request_list)
     if error_check == "Good to go!":
-        return response_ok(request_list[1])
+        body_and_type = resolve_uri(request_list[1])
+        return response_ok(request_list[1], body_and_type)
     return error_check
 
 
 def check_for_errors(request):
+    if len(request) < 5:
+        return response_error('400', 'BAD REQUEST')
     if request[0] != 'GET':
         return response_error('405', '{} METHOD NOT ALLOWED'.format(request[0]))
+    if not os.path.exists(request[1]):
+        return response_error('404', 'NOT FOUND')
     if request[2] != 'HTTP/1.1':
         return response_error('505', '{} NOT SUPPORTED'.format(request[2]))
     return "Good to go!"
 
 
+def resolve_uri(uri):
+    if os.path.isdir(uri):
+        return (generate_dir_html(uri), 'text/html')
+    return (read_file_data(uri), mimetypes.guess_type(uri)[0])
+
+
+def read_file_data(uri):
+    read_data = ""
+    with open(uri, 'r') as f:
+        read_data = f.read()
+    return read_data
+
+
+def generate_dir_html(uri):
+    dirname = "<h1>{}</h1>".format(uri)
+    items_in_dir = "<ul>"
+    for item in os.listdir(uri):
+        items_in_dir = "{}<li>{}</li>".format(items_in_dir, item)
+    items_in_dir = "{}</ul>".format(items_in_dir)
+    return "{}{}".format(dirname, items_in_dir)
```

Comment:
I'm growing less and less fond of building things this way, instead of appending html lines to a list and then joining them.